File: vscode/src/host/localizations/langFileWriter.ts
import * as fs from 'fs';
import * as vscode from 'vscode';
import { withValue } from '../../core/localizations/langJson';

/**
 * Port of `dev.nylo.plugin.localizations.edit.LangFileWriter`. Writes a single translation value back
 * into a `lang/<code>.json` file. Goes through a `WorkspaceEdit` when the file is open (so an editor
 * refreshes and the change joins undo); falls back to a direct write otherwise. Returns null on success.
 */
export async function writeValue(filePath: string, key: string, value: string): Promise<string | null> {
  const open = vscode.workspace.textDocuments.find((d) => d.uri.fsPath === filePath && !d.isClosed);
  try {
    if (open) {
      const newText = withValue(open.getText(), key, value);
      const edit = new vscode.WorkspaceEdit();
      const fullRange = new vscode.Range(open.positionAt(0), open.positionAt(open.getText().length));
      edit.replace(open.uri, fullRange, newText);
      if (!(await vscode.workspace.applyEdit(edit))) {
        return `Could not apply the edit to ${key}.`;
      }
      await open.save();
      return null;
    }
    const text = fs.readFileSync(filePath, 'utf8');
    fs.writeFileSync(filePath, withValue(text, key, value), 'utf8');
    return null;
  } catch (e) {
    return `Failed to write ${key}: ${(e as Error).message}`;
  }
}

File: intellij/src/main/kotlin/dev/nylo/plugin/localizations/edit/LangFileWriter.kt
package dev.nylo.plugin.localizations.edit

import com.intellij.openapi.command.WriteCommandAction
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.LocalFileSystem
import com.intellij.openapi.vfs.VfsUtil
import dev.nylo.plugin.localizations.json.LangJson
import java.io.File

/**
 * Writes a single translation value back into a `lang/<code>.json` file for inline edits.
 *
 * Goes through the IDE document (when one is loaded) so an open editor refreshes and the change joins
 * the undo stack; falls back to a direct VFS write otherwise. Must be called on the EDT — it mutates
 * inside a [WriteCommandAction]. Returns null on success, else a user-facing error message.
 */
object LangFileWriter {
    fun setValue(project: Project, file: File, key: String, value: String): String? {
        val vFile = LocalFileSystem.getInstance().refreshAndFindFileByIoFile(file)
            ?: return "Could not locate ${file.name}"
        var error: String? = null
        WriteCommandAction.runWriteCommandAction(project, "Edit Translation", null, Runnable {
            error = runCatching {
                val fdm = FileDocumentManager.getInstance()
                val document = fdm.getDocument(vFile)
                if (document != null) {
                    document.setText(LangJson.withValue(document.text, key, value))
                    fdm.saveDocument(document)
                } else {
                    VfsUtil.saveText(vFile, LangJson.withValue(VfsUtil.loadText(vFile), key, value))
                }
                // A null message (e.g. an NPE) must still surface as a failure, not read as success.
            }.exceptionOrNull()?.let { it.message ?: it.javaClass.simpleName }
        })
        return error
    }
}
